packages/circuit-ui/components/Theming/Theming.docs.mdx

```diff
@@ -0,0 +1,168 @@
+  };
+};
+type BgNeutrals = {
+  default: ColorWithState;
+  subtle: ColorWithState;
+  highlight: ColorWithState;
+};
+type BgOrBorder = BgRoles & BgNeutrals;
+
+export type Theme = {
+  bg: BgOrBorder;
+  fg: Fg;
+  border: BgOrBorder;
+};
+```
+
+...and as JSON with fake colors, for illustration purposes:
+
+```json
+{
+  "fg": {
+    // neutrals
+    "default": "darkgray",
+    "subtle": "gray",
+    "highlight": "black",
+    // roles (example using primary, same for confirm, notify, alert)
+    "primary": "darkblue",
+    "onPrimary": "white"
+  },
+  "bg": {
+    // neutrals
+    "default": {
+      "idle": "white",
+      "hover": "lightestgray",
+      "active": "lightergray"
+    },
+    "subtle": {
+      "idle": "lightergray",
+      "hover": "lightgray",
+      "active": "gray"
+    },
+    // roles (example using primary, same for confirm, notify, alert)
+    "primary": {
+      "default": {
+        "idle": "darkblue",
+        "hover": "darkerblue",
+        "active": "darkestblue"
+      }
+      // same for other role bg variants, e.g. subtle
+    }
+  },
+  "border": {} // same as bg
+}
+```
+
+### Rationale
+
+- On the top-level, we split foreground colors and background/border colors, because foreground colors have no variants (except neutrals) and states* (*: see open questions)
+- We split neutral colors and role colors, because neutrals typically have more variants (e.g. default neutral text, subtle neutral text, highlight neutral text—but only a single variant of alert text)
+- We split backgrounds and borders because they vary in some cases (e.g. in the light theme, default bg is white and default border is light gray), and also because borders will need to have increased contrast in future high-contrast modes
+
+### Contrast requirements
+
+These can also be programmatically validated in tests.
+
+- Color roles as backgrounds (e.g. `bg.primary.default.*`) need at least 4.5:1 contrast with the accompanying `on*` foreground color (e.g. `fg.onPrimary`)
+- Color roles as foregrounds (e.g. `fg.primary`) need at least 4.5:1 contrast with the default background (`bg.default.*`) (Note: see open question about the notify color role)
+- Subtle variants of color roles (e.g. `bg.primary.subtle.*`) need at least 4.5:1 contrast with neutral fg colors (`fg.default`, `fg.subtle`...)
+- Neutral foregrounds need at least 4.5:1 contrast with neutral backgrounds
+- 🚧
+
+### Open questions
+
+- Can we group text and icon colors under `fg`?
+- Is `promo` another color role, or a different thing?
+- In our current light theme, `fg.notify` (e.g. used in notify icons) doesn't have 4.5:1 contrast to `bg.default`. It would be tricky to change this color—can we rely on documentation _not_ to use notify as text color? Or do we need to split icons from text? (`text.notify` could be set as gray, like in the `Input`'s warning message)
```

Comment:
I'd love that too (very strong aspect of Primer IMO) but it's a heavy brand change. I'll check what we can do with design

Reply:
We discussed that it would be good to see this as an enabling constraint on designs. It doesn't mean that we must change the notify color today, but we can see our current notify color as a bug that we should address moving forward

Reply:
Also, if we decide to go that route, does it make sense to use these constraints:

- `foreground.<role>.<all states>` must have >4.5:1 contrast with `background.default.<all states>` (i.e. this is what we discuss above)
  - e.g. in light mode: dark green text (`foreground.confirm.idle`) on white background (`background.default.idle`) (use case: `Body` with `variant="confirm"`)
  - ⚠️ currently our notify color would fail against `background.default.idle`. We can start the process of finding a better alternative with brand
- `background.<role>.subtle.<all states>` must have >4.5:1 contrast with `foreground.default.<all states>`
  - e.g. in light mode: light blue background (`background.accent.subtle.idle`) and default text (`foreground.default.idle`) (use case: `NotificationBanner`)
- `background.<role>.default.<all states>` must have >4.5:1 contrast with a new `foreground.onRole.<all states>`
  - e.g. in light mode: dark blue background (`background.accent.default.idle`) and white text (`foreground.onRole.idle`)
  - I don't like the term "onRole" much. We could also flip it around and use the "subtle" bg variant as default (the one that works with default text), name the other bg variant "highlight", and use the foreground color name "onHighlight". This would make naming clearer, but it might be strange in our case: since these "subtle" bg variants are practically non-existent today (except the blue in the `NotificationBanner`), using them as default may confuse people.

Following these constraints would get us really close to Primer, especially the last bullet above is what GitHub does, except they name it "emphasis/onEmphasis".

Another benefit is that we wouldn't need to support `on*` colors for each role, if we agree to the constraints: we would simply need to make sure that all `<role>.default` backgrounds work with the `default` foreground, and that all `<role>.highlight` backgrounds work with the `onHighlight` foreground.

What do you think @connor-baer?

Reply:
Also for the record, an alternative to all of this would be the following:

- `icon.notify.idle` could be yellow (slightly darker than the current one) and work with `background.default.idle` (white) (because icons only requires >3:1)
- `text.notify.idle` could be gray and work with `background.default.idle` (white) (like it is currently)
- `background.notify.default.idle` can be yellow (slightly darker than the current one) and work with `text.onNotify.idle` (white)

But on second thought, this is not necessarily better because:

- I thought that our current notify color already had >3:1 with white (works for graphical elements, e.g. borders or icons). It doesn't. So we need to adapt the notify anyways to make it work in notification borders, warning icons, etc.
- generally, in other themes (dark themes), the yellow might have good contrast with the default background—so it makes sense, I think, to treat it the same as other colors. Yellow having low contrast to white doesn't mean it will have low contrast in all situations and be treated differently, or avoided

Reply:
I like those constraints and that they move us closer to Primer's approach. 

> I don't like the term "onRole" much.

What about `onVariant`? Or is the term "variant" reserved for something else?

Reply:
Yes `variant` would be e.g. "subtle" or "highlight" in this case, so I'd avoid mixing terms up here. Maybe `onColor` (basically everything but not neutral, although yes, I know, black and white are colors). Or `onColorRole`? Or we just call the dark blue/green/red "highlight" and call it `onHighlight`, like Primer does.

Reply:
Ah, I got confused because of "primary" which is a button variant and was a "role" until we renamed it to accent (good decision!).

Hm, I prefer "onHighlight" to "onRole" though neither makes me truly happy 😄